Added user defined custom task content (#84)
osrf/ros2multirobotbook: src/task_userdefined.md

# User-defined Custom Tasks in RMF Task

> >Note: User-defined custom tasks are currently experimental

Reviewer: Let's also add that there is a new general task composition system in coming and it's under discussion here: https://github.com/open-rmf/rmf/discussions/169

When dealing with rmf_task, there are two packages. 
- rmf_task 

> - rmf_task_sequence

Reviewer: Link the pacakges?

> `rmf_task` provides APIs and base classes for defining and managing Tasks in RMF. A Task is defined as an object that generates phases which are a meaningful sequence of steps that results in a desirable outcome.  Each task will have a description and a component that allows us to model the state of the robot after completing the task given its initial state and also a component that will command the actual robot to perform the task.

Reviewer: ```suggestion
`rmf_task` provides APIs and base classes for defining and managing Tasks in RMF. A Task is defined as an object that generates phases which are a meaningful sequence of steps that results in a desirable outcome. Each task will have a description and a component that allows us to model the state of the robot after completing the task given its initial state and also a component that will command the actual robot to perform the task.
```

> `rmf_task_sequence` provides an out of the box implementation of rmf_task where a Task object is defined by a sequence of phases. The phases that such tasks generate will thus match the sequence of phases used to define them. 

Reviewer: ```suggestion
`rmf_task_sequence` provides an out of the box implementation of `rmf_task` where a `Task` object is defined by a sequence of phases. The phases that such tasks generate will thus match the sequence of phases used to define them. 
```

> The phases defined in rmf_task_sequence are in turn a collection of events which also have components to model the end state and command the robot during the event. Presently the following events are supported  are defined [here](https://github.com/open-rmf/rmf_task/tree/main/rmf_task_sequence/include/rmf_task_sequence/events)

Reviewer: ```suggestion
The phases defined in `rmf_task_sequence` are in turn a collection of events which also have components to model the end state and command the robot during the event. Presently the following events are supported and defined [here](https://github.com/open-rmf/rmf_task/tree/main/rmf_task_sequence/include/rmf_task_sequence/events).
```

Users can then construct arbitrary definitions of tasks by stringing together a sequence of such phases/events. RMF is capable of planning and executing such tasks.

> Users will be working on `rmf_fleet_adapter` layer. In this layer API is restricted to use only `rmf_task_sequence` to perform tasks and only certain events are supported their description can be found [here](https://github.com/open-rmf/rmf_ros2/tree/main/rmf_fleet_adapter/schemas)

Reviewer: ```suggestion
Users will be working on the `rmf_fleet_adapter` layer. In this layer the API is restricted to use only `rmf_task_sequence` to perform tasks. Only certain events are supported, their description can be found [here](https://github.com/open-rmf/rmf_ros2/tree/main/rmf_fleet_adapter/schemas).
```

> The `rmf_fleet_adapter` layer acts as an API that users can use and it supports customised behaviour on only `perform_action` in already existing  events mentioned [here](https://github.com/open-rmf/rmf_ros2/tree/main/rmf_fleet_adapter/schemas) .

Reviewer: ```suggestion
The `rmf_fleet_adapter` layer acts as an API that users can use. It supports customised behaviour on only `perform_action` in the already existing events mentioned [here](https://github.com/open-rmf/rmf_ros2/tree/main/rmf_fleet_adapter/schemas).
```

Reviewer: Also, can you add some explanation on what is this `perform_action`?

> Users can only add custom tasks in `perform_action`. RMF passes the command to the platform-specific end of the fleet adapter integration and formally releases control of the bot until the action is finished.

Reviewer: ```suggestion
Users can only add custom tasks in `perform_action`. RMF passes the command to the platform-specific end of the fleet adapter integration and formally releases control of the robot until the action is finished.
```

To use a custom task in `perform_action` users need to use two parts of the API. 
1. [FleetUpdateHandle::add_performable_action](https://github.com/open-rmf/rmf_ros2/blob/8440488d5583edc5a5b7226326aa2a8d41dad975/rmf_fleet_adapter/include/rmf_fleet_adapter/agv/FleetUpdateHandle.hpp#L188-L201)

> This consists of two things one is the `category` of the action and the second is the `consider` based on which it would be decided whether to accept the action or not.

Reviewer: ```suggestion
This consists of two parts: the first one is the `category` of the action and the second is the `consider` part, based on which it would be decided whether to accept the action or not.
```

> example

Reviewer: ```suggestion
Here is an example:
```

    delivery: False
    clean: True
    finishing_request: "park"
    action_categories: ["clean", "manual_control"]

```

```python
def _consider(description: dict):
        confirm = adpt.fleet_update_handle.Confirmation()

        # Currently there's no way for user to submit a robot_task_request
        # .json file via the rmf-web dashboard. Thus the short term solution
        # is to add the fleet_name info into action description. NOTE
        # only matching fleet_name action will get accepted
        if (description["category"] == "manual_control" and
            description["description"]["fleet_name"] != fleet_name):
                return confirm

        node.get_logger().warn(
            f"Accepting action: {description} ")
        confirm.accept()
        return confirm
fleet_config = config_yaml['rmf_fleet']
task_capabilities_config = fleet_config['task_capabilities']
if 'action_categories' in task_capabilities_config:

	for cat in task_capabilities_config['action_categories']:
	
		node.get_logger().info(
		
			f"Fleet [{fleet_name}] is configured"
			
			f" to perform action of category [{cat}]")
	
		fleet_handle.add_performable_action(cat, _consider)
```

> 2. [RobotUpdateHandle::set_action_executor](https://github.com/open-rmf/rmf_ros2/blob/8440488d5583edc5a5b7226326aa2a8d41dad975/rmf_fleet_adapter/include/rmf_fleet_adapter/agv/RobotUpdateHandle.hpp#L178-L179) This is where you tell the fleet adapter how to instruct your robot to begin performing an action. This callback to this function consists of 

Reviewer: ```suggestion
2. [RobotUpdateHandle::set_action_executor](https://github.com/open-rmf/rmf_ros2/blob/8440488d5583edc5a5b7226326aa2a8d41dad975/rmf_fleet_adapter/include/rmf_fleet_adapter/agv/RobotUpdateHandle.hpp#L178-L179) This is where you tell the fleet adapter how to instruct your robot to begin performing an action. This callback to this function consists of:
```

> - `category(string)`  type of action

Reviewer: ```suggestion
- `category(string)`  type of action.
```

> - `description(JSON)`  message which contains details about how the action should be performed

Reviewer: ```suggestion
- `description(JSON)`  message which contains details about how the action should be performed.
```

- `execution(object)` object which the platform-specific side of the fleet adapter must hold onto while the action is being performed, ideally giving periodic updates for remaining time estimates.

> The robot will not participate in the traffic negotiations while using a custom task in perform_action. That means that the robot will be allowed to report its trajectory to the traffic schedule thus making it possible for other robots to avoid the bot however the bot would not be able to accommodate other bots until the task is complete.

Reviewer: ```suggestion
The robot will not participate in the traffic negotiations while using a custom task in `perform_action`. That means that it will be allowed to report its trajectory to the traffic schedule, thus making it possible for other robots to avoid it. However, the robot would not be able to accommodate other robots until the task is complete.
```

Reviewer: Let's not call it `bot`, I believe that term is mostly used for software programs.

> example

Reviewer: ```suggestion
Here is an example:
```

                            f"Failed to initiate cleaning action for robot [{self.name}]")
                        # TODO: issue error ticket
                        self.in_error = True # TODO: toggle error back
                        execution.error("Failed to initiate cleaning action for robot {self.name}")
                        execution.finished()
                        return
                    attempts+=1
                    time.sleep(1.0)
            elif (category == "manual_control"):
                self.check_task_completion = lambda:False  # user can only cancel the manual_control

            # start Perform Action
            self.node.get_logger().warn(f"Robot [{self.name}] starts [{category}] action")
            self.start_action_time = self.adapter.now()
            self.on_waypoint = None
            self.on_lane = None
            self.action_execution = execution
            self.stubbornness = self.update_handle.unstable_be_stubborn()
            # robot moves slower during perform action
            self.vehicle_traits.linear.nominal_velocity *= 0.2
```

```python
self.update_handle.set_action_executor(self._action_executor)
```

> The examples are part of the following  [repository](https://github.com/open-rmf/fleet_adapter_ecobot/tree/70dab6e657ca281bb0a299b7fe03785d4d4e1d54).

Reviewer: ```suggestion
These examples are part of the following [repository](https://github.com/open-rmf/fleet_adapter_ecobot/tree/70dab6e657ca281bb0a299b7fe03785d4d4e1d54).
```

# User-defined Custom Tasks in RMF Task

> Note: User-defined custom tasks are currently experimental

> A new general task composition system is under progress the discussion can be found [here](https://github.com/open-rmf/rmf/discussions/169).

> When dealing with rmf_task, there are two packages.

Reviewer: ```suggestion
When dealing with RMF tasks, there are two packages:
```

- [rmf_task](https://github.com/open-rmf/rmf_task/tree/main/rmf_task)
- [rmf_task_sequence](https://github.com/open-rmf/rmf_task/tree/main/rmf_task_sequence)

`rmf_task` provides APIs and base classes for defining and managing Tasks in RMF. A Task is defined as an object that generates phases which are a meaningful sequence of steps that results in a desirable outcome. Each task will have a description and a component that allows us to model the state of the robot after completing the task given its initial state and also a component that will command the actual robot to perform the task.

`rmf_task_sequence` provides an out of the box implementation of `rmf_task` where a `Task` object is defined by a sequence of phases. The phases that such tasks generate will thus match the sequence of phases used to define them.

> The phases defined in `rmf_task_sequence` are in turn a collection of events which also have components to model the end state and command the robot during the event. Presently the following events are supported and defined [here](https://github.com/open-rmf/rmf_task/tree/main/rmf_task_sequence/include/rmf_task_sequence/events).

Reviewer: ```suggestion
The phases defined in `rmf_task_sequence` are in turn a collection of events which also have components to model the end state and command the robot during the event. Presently the events defined [here](https://github.com/open-rmf/rmf_task/tree/main/rmf_task_sequence/include/rmf_task_sequence/events) are supported.
```

    delivery: False
    clean: True
    finishing_request: "park"
    action_categories: ["clean", "manual_control"]
```

```python
def _consider(description: dict):
        confirm = adpt.fleet_update_handle.Confirmation()

        # Currently there's no way for user to submit a robot_task_request
        # .json file via the rmf-web dashboard. Thus the short term solution
        # is to add the fleet_name info into action description. NOTE
        # only matching fleet_name action will get accepted
        if (description["category"] == "manual_control" and
            description["description"]["fleet_name"] != fleet_name):
                return confirm

        node.get_logger().warn(
            f"Accepting action: {description} ")
        confirm.accept()
        return confirm
fleet_config = config_yaml['rmf_fleet']
task_capabilities_config = fleet_config['task_capabilities']
if 'action_categories' in task_capabilities_config:

	for cat in task_capabilities_config['action_categories']:

		node.get_logger().info(

			f"Fleet [{fleet_name}] is configured"

			f" to perform action of category [{cat}]")

		fleet_handle.add_performable_action(cat, _consider)
```

> 2. [RobotUpdateHandle::set_action_executor](https://github.com/open-rmf/rmf_ros2/blob/8440488d5583edc5a5b7226326aa2a8d41dad975/rmf_fleet_adapter/include/rmf_fleet_adapter/agv/RobotUpdateHandle.hpp#L178-L179) This is where you tell the fleet adapter how to instruct your robot to begin performing an action. This callback to this function consists of:

Reviewer: ```suggestion
2. [RobotUpdateHandle::set_action_executor](https://github.com/open-rmf/rmf_ros2/blob/8440488d5583edc5a5b7226326aa2a8d41dad975/rmf_fleet_adapter/include/rmf_fleet_adapter/agv/RobotUpdateHandle.hpp#L178-L179) This is where you tell the fleet adapter how to instruct your robot to begin performing an action. The callback on this function consists of:
```

- `category(string)` type of action.

> - `description(JSON)`  message which contains details about how the action should be performed.
> - `execution(object)` object which the platform-specific side of the fleet adapter must hold onto while the action is being performed, ideally giving periodic updates for remaining time estimates.

Reviewer: ```suggestion
- `description(JSON)` message which contains details about how the action should be performed.
- `execution(object)`object which the platform-specific side of the fleet adapter must hold onto while the action is being performed, ideally giving periodic updates for remaining time estimates.
```
